## ID : 816000325
## Name: Ajay Sieunarine
## Email: ajay.sieunarine@my.uwi.edu

# Part A

## 1. Convert the following sentence to a python tuple list of letters and their frequency as they appear. Ignore all non-alpha numeric characters.

Sentence: 

“The quick brown fox jumps over the lazy dog and the fox was very happy”

Tuple List:

[('h', 1), ('e', 1), ('t', 1), ('q', 1), ('i', 1), ('c', 1), ('u', 1), ('k', 1), ('r', 1), ('b', 1), ('o', 1), ('w', 1), ('n', 1), ('x', 1), ('o', 1), ('f', 1), ('u', 1), ('s', 1), ('j', 1), ('m', 1), ('p', 1), ('r', 1), ('e', 1), ('o', 1), ('v', 1), ('h', 1), ('e', 1), ('t', 1), ('a', 1), ('y', 1), ('z', 1), ('l', 1), ('o', 1), ('d', 1), ('g', 1), ('a', 1), ('d', 1), ('n', 1), ('h', 1), ('e', 1), ('t', 1), ('x', 1), ('o', 1), ('f',1), ('a', 1), ('s', 1), ('w', 1), ('y', 1), ('r', 1), ('e', 1), ('v', 1), ('a', 1), ('h', 1), ('y', 1), ('p', 2)]



In [529]:
foxy_sentence = "The quick brown fox jumps over the lazy dog and the fox was very happy"
new_sentence = "" # empty string which will represent the sentence with all common and letters only (with spaces)

for i in foxy_sentence: # populate new sentence
    if i.isalpha() or i == " ":
        new_sentence += i.lower()

# print (new_sentence)
words = new_sentence.split() # words will be an array of strings representing all the words in the new_sentence
# print(words)

foxy_tuple = [] # the tuple to hold the date

# foreach word, grab the frequency of each letter, 
# and store it in a new tuple, which will then be appended to the actual tuple
for word in words: 
    freqs = {}
    for ch in word: # 
        if ch in freqs:
            freqs[ch] += 1
        else:
            freqs[ch] = 1
    temp_tuple = tuple(freqs.items())
    foxy_tuple += temp_tuple


print(type(foxy_tuple))
print(len(foxy_tuple))
print(foxy_tuple)

[('t', 1), ('h', 1), ('e', 1), ('q', 1), ('u', 1), ('i', 1), ('c', 1), ('k', 1), ('b', 1), ('r', 1), ('o', 1), ('w', 1), ('n', 1), ('f', 1), ('o', 1), ('x', 1), ('j', 1), ('u', 1), ('m', 1), ('p', 1), ('s', 1), ('o', 1), ('v', 1), ('e', 1), ('r', 1), ('t', 1), ('h', 1), ('e', 1), ('l', 1), ('a', 1), ('z', 1), ('y', 1), ('d', 1), ('o', 1), ('g', 1), ('a', 1), ('n', 1), ('d', 1), ('t', 1), ('h', 1), ('e', 1), ('f', 1), ('o', 1), ('x', 1), ('w', 1), ('a', 1), ('s', 1), ('v', 1), ('e', 1), ('r', 1), ('y', 1), ('h', 1), ('a', 1), ('p', 2), ('y', 1)]
<class 'list'>
55


## 2. Create a PySpark Context

In [530]:
import pyspark # library for the spark session -> spark will allow us to manipualte RDDs
print ("imported pyspark")
from pyspark.sql import SparkSession # spark session -> allow for dataframe and sql functionality
from pyspark.sql import Row # allow us to represent rows from dataframes
from pyspark.sql.types import * # get/set datatypes of cols
from pyspark.sql.functions import * # epic built-in functions 
from pyspark.ml.linalg import DenseVector # for linear algebra, uses numpy arrays,
from pyspark.ml.feature import StandardScaler # standardize features by removing the mean, and scaling to unit variance
print("imported all pyspark libs")
import pprint

# init spark sesh
spark = SparkSession.builder \
    .master("local") \
    .appName("Assignment 2") \
    .config("spark.executor.memory", "1gb") \
    .getOrCreate()

print ("initialized spark session")

sc = spark.sparkContext
print(sc)

imported pyspark
imported all pyspark libs
initialized spark session
<SparkContext master=local appName=Assignment 2>


## 3. Convert the list of tuples to a PySpark RDD. 

In [552]:
rdd = sc.parallelize(foxy_tuple) # make rdd from tuple
type(rdd)
rdd_list = rdd.collect()
for el in rdd_list:
    print(el)

('t', 1)
('h', 1)
('e', 1)
('q', 1)
('u', 1)
('i', 1)
('c', 1)
('k', 1)
('b', 1)
('r', 1)
('o', 1)
('w', 1)
('n', 1)
('f', 1)
('o', 1)
('x', 1)
('j', 1)
('u', 1)
('m', 1)
('p', 1)
('s', 1)
('o', 1)
('v', 1)
('e', 1)
('r', 1)
('t', 1)
('h', 1)
('e', 1)
('l', 1)
('a', 1)
('z', 1)
('y', 1)
('d', 1)
('o', 1)
('g', 1)
('a', 1)
('n', 1)
('d', 1)
('t', 1)
('h', 1)
('e', 1)
('f', 1)
('o', 1)
('x', 1)
('w', 1)
('a', 1)
('s', 1)
('v', 1)
('e', 1)
('r', 1)
('y', 1)
('h', 1)
('a', 1)
('p', 2)
('y', 1)


## 4. Using the methods of PySpark RDD display the letter count.

In [553]:
rdd_grouped = rdd.reduceByKey(lambda x, y: x+y)
rdd_grouped = rdd_grouped.sortByKey()
rdd_letter_count = rdd_grouped.collect()
for el in rdd_letter_count:
    print(el)

('a', 4)
('b', 1)
('c', 1)
('d', 2)
('e', 5)
('f', 2)
('g', 1)
('h', 4)
('i', 1)
('j', 1)
('k', 1)
('l', 1)
('m', 1)
('n', 2)
('o', 5)
('p', 3)
('q', 1)
('r', 3)
('s', 2)
('t', 3)
('u', 2)
('v', 2)
('w', 2)
('x', 2)
('y', 3)
('z', 1)


## 5. Using the methods of PySpark RDD display the letter and number of times they appear in each word in the sentence.

Sample:

[('a', [1, 1, 1, 1]), ('e', [1, 1, 1, 1, 1]), ('i', [1]), ('m', [1])]

In [554]:
rdd_group_by_key = rdd.groupByKey() # make each entry in the rdd iterable
print(type(rdd))
    
rdd_word_occurences = rdd_group_by_key.map(
    lambda x: (
        x[0], # the letter
        list(x[1]) # the occurences as a list
    )
)
rdd_word_occurences = rdd_word_occurences.sortByKey() # sort in alphabetical order
rdd_list = rdd_word_occurences.collect()
for el in rdd_list:
    print(el)

<class 'pyspark.rdd.RDD'>
('a', [1, 1, 1, 1])
('b', [1])
('c', [1])
('d', [1, 1])
('e', [1, 1, 1, 1, 1])
('f', [1, 1])
('g', [1])
('h', [1, 1, 1, 1])
('i', [1])
('j', [1])
('k', [1])
('l', [1])
('m', [1])
('n', [1, 1])
('o', [1, 1, 1, 1, 1])
('p', [1, 2])
('q', [1])
('r', [1, 1, 1])
('s', [1, 1])
('t', [1, 1, 1])
('u', [1, 1])
('v', [1, 1])
('w', [1, 1])
('x', [1, 1])
('y', [1, 1, 1])
('z', [1])


# Part B

If you are a frequent user of Amazon.com, you are probably familiar with the lists of related products (books, CDs, etc.) the site features to help customers find what they are looking for. Amazon.com presents several such lists on every page, including “Frequently Bought Together” and “Customers Who Bought This Item Also Bought.” These features have roots and solutions in recommendation engines and systems. 

## Determine:
•	Customers who bought this item also bought  
•	Most popular items 


## 1.	Create a sql context from PySpark SQLContext.  	 	

In [534]:
from pyspark.sql import SQLContext

sql = SQLContext(sc)
print(sql)

## 2.	Load the Amazon Review Dataset into a PySpark RDD, ensure that each row is properly separated and the headers are matched to their respective columns.  

In [535]:
rdd = sc.textFile('amazon_reviews.csv')
rdd.collect()
print(rdd)

amazon_reviews.csv MapPartitionsRDD[1987] at textFile at <unknown>:0


In [536]:
print(rdd.take(5))

[',id,name,username', '0,AVpe7AsMilAPnD_xQ78G,Kindle Paperwhite,Cristina M', '1,AVpe7AsMilAPnD_xQ78G,Kindle Paperwhite,Ricky', '2,AVpe7AsMilAPnD_xQ78G,Kindle Paperwhite,Tedd Gardiner', '3,AVpe7AsMilAPnD_xQ78G,Kindle Paperwhite,Dougal']


In [537]:
rdd = rdd.map(lambda line: line.split(",")) # split by delim=','

In [538]:
rdd.take(5) # if getting split not in list error, run all cells

[['', 'id', 'name', 'username'],
 ['0', 'AVpe7AsMilAPnD_xQ78G', 'Kindle Paperwhite', 'Cristina M'],
 ['1', 'AVpe7AsMilAPnD_xQ78G', 'Kindle Paperwhite', 'Ricky'],
 ['2', 'AVpe7AsMilAPnD_xQ78G', 'Kindle Paperwhite', 'Tedd Gardiner'],
 ['3', 'AVpe7AsMilAPnD_xQ78G', 'Kindle Paperwhite', 'Dougal']]

### Cleaning some string records by removing non-alphabetical characters.

In [539]:
# def cleanName(x):
#     newName = ""
#     for name in x:
#         for ch in name:
#             if(ch.isalpha() or ch.isdigit() or ch == ' '):
#                 newName += ch
#     return newName
                

# rdd_cleaned = rdd.map(
#     lambda x: (
#         x[0], x[1], cleanName(x[2]), cleanName(x[3])
#     )
# )

# rdd = rdd_cleaned

## 3.	Convert the rdd to a PySpark DataFrame. 

In [540]:
print("mapping rdd into df...")
df = rdd.map(lambda line: Row(
    key=line[0],
    itemID=line[1],
    itemName=line[2],
    username=line[3],
)).toDF()
df = df[['key', 'itemID', 'itemName', 'username']]
# df.drop([0,1,2])
df = df.where("itemName!='name'") # remove header row
print("done.")
print(type(df))

mapping rdd into df...
done.
<class 'pyspark.sql.dataframe.DataFrame'>


In [541]:
df.show(3)

+---+--------------------+-----------------+-------------+
|key|              itemID|         itemName|     username|
+---+--------------------+-----------------+-------------+
|  0|AVpe7AsMilAPnD_xQ78G|Kindle Paperwhite|   Cristina M|
|  1|AVpe7AsMilAPnD_xQ78G|Kindle Paperwhite|        Ricky|
|  2|AVpe7AsMilAPnD_xQ78G|Kindle Paperwhite|Tedd Gardiner|
+---+--------------------+-----------------+-------------+
only showing top 3 rows



In [542]:
df.printSchema()

root
 |-- key: string (nullable = true)
 |-- itemID: string (nullable = true)
 |-- itemName: string (nullable = true)
 |-- username: string (nullable = true)



## 4.	Using the dataframe from the above question show the top 20 bought products

In [543]:
mostBought = df.groupBy("itemName").count()
print(type(mostBought))
mostBought.sort("count", ascending=False).show(20)

<class 'pyspark.sql.dataframe.DataFrame'>
+--------------------+-----+
|            itemName|count|
+--------------------+-----+
|Amazon Tap - Alex...|  542|
|      Amazon Fire TV|  166|
|Amazon Premium He...|  133|
|    Fire HD 6 Tablet|   87|
|"Kindle Fire HDX ...|   53|
|"Kindle Fire HDX ...|   43|
|"Kindle Fire HD 7"""|   41|
|   Kindle Paperwhite|   39|
|Certified Refurbi...|   38|
|     Kindle Keyboard|   32|
|All-New Amazon Fi...|   27|
|              Kindle|   20|
|Amazon 5W USB Off...|   19|
|All-New Amazon Fi...|   18|
|Replacement Remot...|   17|
|Echo Dot (2nd Gen...|   13|
|All-New Amazon Ki...|   13|
|Moshi Anti-Glare ...|   12|
|All-New Amazon Ki...|   12|
|Alexa Voice Remot...|   12|
+--------------------+-----+
only showing top 20 rows



## 5.	Using the datafrme from the previous question show the top 20 users and the products that they purchased. 

In [544]:
bestCustomers = df.groupBy("username", "itemName").count()
bestCustomers = bestCustomers.sort("count", ascending=False)
bestCustomers.show(20)

+-------------------+--------------------+-----+
|           username|            itemName|count|
+-------------------+--------------------+-----+
|          A. Younan|Amazon Premium He...|   59|
|             Andrew|Amazon Premium He...|   43|
|          Victor L.|Amazon Premium He...|   30|
|     William Hardin|    Fire HD 6 Tablet|   30|
|            Mike W.|    Fire HD 6 Tablet|   29|
|      Earthling1984|    Fire HD 6 Tablet|   28|
|     William Hardin|      Amazon Fire TV|   16|
|         B. Tarbuck|"Kindle Fire HDX ...|   16|
|              Mandy|      Amazon Fire TV|   16|
|                 NF|"Kindle Fire HDX ...|   15|
|                 NF|"Kindle Fire HD 7"""|   14|
|    Amazon Reviewer|"Kindle Fire HDX ...|   14|
|    Amazon Reviewer|"Kindle Fire HDX ...|   13|
|             Dallas|      Amazon Fire TV|   12|
|     William Hardin|Certified Refurbi...|   12|
|  Michael Gallagher|"Kindle Fire HDX ...|   12|
|   Gregory P. Baker|      Amazon Fire TV|   11|
|              MarkM

## 6.	Create a RDD of tuples from the dataframe with only 2 columns ‘product’ and ‘username’ in that order.         

In [545]:
df1 = df.select("itemName", "username") # df with products and username
df1 = df1.selectExpr("itemName as product", "username as username") # rename col
df1.show(5)

+-----------------+------------------+
|          product|          username|
+-----------------+------------------+
|Kindle Paperwhite|        Cristina M|
|Kindle Paperwhite|             Ricky|
|Kindle Paperwhite|     Tedd Gardiner|
|Kindle Paperwhite|            Dougal|
|Kindle Paperwhite|Miljan David Tanic|
+-----------------+------------------+
only showing top 5 rows



In [546]:
rdd1 = df1.rdd.map(lambda x: tuple(x))
print(rdd1)
print(rdd1.count())
print(rdd1.take(3))

PythonRDD[2022] at RDD at PythonRDD.scala:53
1580
[('Kindle Paperwhite', 'Cristina M'), ('Kindle Paperwhite', 'Ricky'), ('Kindle Paperwhite', 'Tedd Gardiner')]


## 7.	Using methods from PySpark’s RDD object e.g. groupByKey, map, reduceByKey derive the top 20 products. 

Sample: 

[(u'Amazon Tap - Alexa-Enabled Portable Bluetooth Speaker', 542), 
 (u'Amazon Fire TV', 166), 
 (u'Amazon Premium Headphones', 133), 
 (u'Fire HD 6 Tablet', 87), 
 (u'Kindle Fire HDX 7', 53)] 


In [547]:
def countCust(x):
    return len(x)
        
rdd1_grouped = rdd1.groupByKey()

rdd1_key_counts = rdd1_grouped.map(
    lambda x: (
        x[0], countCust(x[1])
    )
)

best_products_list = tuple(rdd1_key_counts.takeOrdered(rdd1_key_counts.count(), key = lambda x: -x[1]))
query_best_prodcuts = sc.parallelize(best_products_list)
products_list = query_best_prodcuts.take(20)
for el in products_list:
    print(el, '\n')

('Amazon Tap - Alexa-Enabled Portable Bluetooth Speaker', 542) 

('Amazon Fire TV', 166) 

('Amazon Premium Headphones', 133) 

('Fire HD 6 Tablet', 87) 

('"Kindle Fire HDX 7"""', 53) 

('"Kindle Fire HDX 8.9"""', 43) 

('"Kindle Fire HD 7"""', 41) 

('Kindle Paperwhite', 39) 

('Certified Refurbished Amazon Fire TV (Previous Generation - 1st)', 38) 

('Kindle Keyboard', 32) 

('All-New Amazon Fire 7 Tablet Case (7th Generation', 27) 

('Kindle', 20) 

('Amazon 5W USB Official OEM Charger and Power Adapter for Fire Tablets and Kindle eReaders', 19) 

('All-New Amazon Fire HD 8 Tablet Case (7th Generation', 18) 

('Replacement Remote for Amazon Fire TV Stick', 17) 

('All-New Amazon Kid-Proof Case for Amazon Fire 7 Tablet (7th Generation', 13) 

('Echo Dot (2nd Generation) - Black', 13) 

('All-New Amazon Kid-Proof Case for Amazon Fire HD 8 Tablet (7th Generation', 12) 

('Moshi Anti-Glare No Bubble Screen Protector for the Fire Phone', 12) 

('Amazon Kindle Oasis Premium Leather Batte

## 8.	Create another RDD of tuples from the dataframe with the columns ‘username’ and ‘product’ in that order.         

In [548]:
df1 = df.select("username", "itemName") # df with products and username
df1 = df1.selectExpr("username as username", "itemName as product") # rename col
df1.show(5)

+------------------+-----------------+
|          username|          product|
+------------------+-----------------+
|        Cristina M|Kindle Paperwhite|
|             Ricky|Kindle Paperwhite|
|     Tedd Gardiner|Kindle Paperwhite|
|            Dougal|Kindle Paperwhite|
|Miljan David Tanic|Kindle Paperwhite|
+------------------+-----------------+
only showing top 5 rows



In [549]:
rdd2 = df1.rdd.map(lambda x: tuple(x))
print(rdd2)
print(rdd2.count())
print(rdd2.take(3))

PythonRDD[2042] at RDD at PythonRDD.scala:53
1580
[('Cristina M', 'Kindle Paperwhite'), ('Ricky', 'Kindle Paperwhite'), ('Tedd Gardiner', 'Kindle Paperwhite')]


## 9. Using methods form PySparks’s RDD object product the top 10 customers who purchased the most items. The top 10 list must show the username and a list of all the items they bought with the number of that item they bought. 

Sample: 

[(u'A. Younan', ({u'Amazon Premium Headphones': 59}, 59)), 
 (u'William Hardin',   ({u'Amazon Fire TV': 16,     u'Certified Refurbished Amazon Fire TV (Previous Generation - 1st)': 12,     u'Fire HD 6 Tablet': 30}, 
   58)), 
 (u'Andrew', ({u'Amazon Premium Headphones': 43}, 43))] 


In [550]:
rdd2_grouped = rdd2.groupByKey()
# print(rdd2_grouped.take(3))

rdd2_products = rdd2_grouped.map(
    lambda x: (
        x[0], list(x[1])
    )
)

rdd2_products = rdd2_products.sortByKey()

# print(rdd2_products.take(2))

def getDict(x):
    freqs = {}
    for item in x:
        if item in freqs:
            freqs[item] += 1
        else:
            freqs[item] = 1
    return freqs

rdd2_products_count = rdd2_products.map(
    lambda x: (
        x[0], getDict(x[1])
    )
)

# print("\n")
# print(rdd2_products_count.take(2))

def getTotal(x):
    total = 0
    index = 0
    for item in x:
        total += x[item]
        index += 1
    return total

rdd2_products_count_total = rdd2_products_count.map(
    lambda x: (
        x[0], x[1], getTotal(x[1])
    )
)

# print("\n")
# print(rdd2_products_count_total.take(5))

print("\n")
best_cust_list = tuple(rdd2_products_count_total.takeOrdered(rdd2_products_count_total.count(), key = lambda x: -x[2]))
query_best_cust = sc.parallelize(best_cust_list)
# pprint(query_best_cust.take(10))
cust_list = query_best_cust.take(20)
for el in cust_list:
    print(el , "\n")




('A. Younan', {'Amazon Premium Headphones': 59}, 59) 

('William Hardin', {'Certified Refurbished Amazon Fire TV (Previous Generation - 1st)': 12, 'Amazon Fire TV': 16, 'Fire HD 6 Tablet': 30}, 58) 

('Andrew', {'Amazon Premium Headphones': 43}, 43) 

('Victor L.', {'Amazon Premium Headphones': 30}, 30) 

('Mike W.', {'Fire HD 6 Tablet': 29}, 29) 

('NF', {'"Kindle Fire HD 7"""': 14, '"Kindle Fire HDX 7"""': 15}, 29) 

('Earthling1984', {'Fire HD 6 Tablet': 28}, 28) 

('Amazon Reviewer', {'"Kindle Fire HDX 8.9"""': 14, '"Kindle Fire HDX 7"""': 13}, 27) 

('Mandy', {'Certified Refurbished Amazon Fire TV (Previous Generation - 1st)': 10, 'Amazon Fire TV': 16}, 26) 

('Amazon Customer', {'"Kindle Fire HDX 8.9"""': 1, 'Certified Refurbished Fire HD 10 Tablet': 1, 'Certified Refurbished Kindle E-reader - Black': 1, 'Certified Refurbished Echo Dot (2nd Generation) - Black': 1, 'Amazon Fire TV': 3, 'Amazon 5W USB Official OEM Charger and Power Adapter for Fire Tablets and Kindle eReaders': 